solr/core/src/java/org/apache/solr/search/SolrIndexSearcher.java

```diff
@@ -1384,16 +1385,24 @@ private void getDocListC(QueryResult qr, QueryCommand cmd) throws IOException {
     // - we don't want score returned.
 
     // check if we should try and use the filter cache
-    boolean useFilterCache = false;
-    if ((flags & (GET_SCORES | NO_CHECK_FILTERCACHE)) == 0 && useFilterForSortedQuery && cmd.getSort() != null
-        && filterCache != null) {
-      useFilterCache = true;
-      SortField[] sfields = cmd.getSort().getSort();
-      for (SortField sf : sfields) {
-        if (sf.getType() == SortField.Type.SCORE) {
-          useFilterCache = false;
-          break;
-        }
+    final boolean needSort;
+    final boolean useFilterCache;
+    if ((flags & (GET_SCORES | NO_CHECK_FILTERCACHE)) != 0 || filterCache == null) {
+      needSort = true; // this value should be irrelevant when `useFilterCache=false`
+      useFilterCache = false;
+    } else {
+      final Sort sort;
+      if (q instanceof  MatchAllDocsQuery // special-case MatchAllDocsQuery: implicit default useFilterForSortedQuery=true
+              || (useFilterForSortedQuery && q instanceof ConstantScoreQuery)) { // default behavior should not risk filterCache thrashing
+        final SortField[] sortFields;
+        // We only need to sort if we're returning results AND sorting by something other than SCORE (sort by
+        // "score" alone is pointless for these constant score queries)
+        needSort = cmd.getLen() > 0 && (sort = cmd.getSort()) != null && ((sortFields = sort.getSort()).length > 1 || sortFields[0].getType() != Type.SCORE);
```

Comment:
This complex conditional with nested assignments is fairly difficult to read, can we expand to something less terse?

Reply:
:+1: ... refactored this (and a couple of other) complex 1-liner conditionals into methods with more transparent names.